server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java

```diff
@@ -35,12 +35,12 @@ private BlobPath(List<String> paths) {
     }
 
     public BlobPath add(String path) {
-        return new BlobPath(CollectionUtils.appendToCopy(this.paths, path));
+        return new BlobPath(CollectionUtils.appendToCopy(this.paths, path.endsWith("/") ? path.substring(0, path.length() - 1) : path));
     }
 
     public String buildAsString() {
         String p = String.join(SEPARATOR, paths);
-        if (p.isEmpty() || p.endsWith(SEPARATOR)) {
+        if (p.isEmpty()) {
```

Comment:
I think if we strip the separator in the only construction path via `add` and stop normalizing here, we get consistent behavior without breaking BwC.
Currently, the fact that we strip the trailing slash leads to broken sub-path building but healthy root path building. This PR does not change the behavior in terms of how root paths are built but makes sure that the same root path (string) will have the same sub paths derived from it.